iree/compiler/Conversion/LinalgToSPIRV/ConvertToSPIRVPass.cpp

```diff
@@ -128,10 +128,13 @@ Value getPushConstantValue(Operation *op, unsigned elementCount,
 /// Gets or inserts a resource evariable of the given `type` in `block` and bind
 /// it to `set` and `binding`.
 spirv::GlobalVariableOp getOrInsertResourceVariable(Location loc, Type type,
+                                                    Type oldType,
                                                     unsigned set,
                                                     unsigned binding,
                                                     Block &block) {
-  auto name = llvm::formatv("__resource_var_{0}_{1}__", set, binding).str();
+  auto name =
+      llvm::formatv("__resource_var_{0}_{1}_{2}__", set, binding, oldType)
+          .str();
```

Comment:
any comment on variable name in getOrInsertResourceVariable would be great... The type name is too long so I use oldType here.

Reply:
Using the typename is strange. I am surprised it round-trips. Maybe just use the function name from where it is accessed.

Reply:
I'm a bit afraid that if this will cause unnecessary memory allocation. According to the definition of globalVariable:

> Allocate an object in memory at module scope. The object is referenced using a symbol name.

We probably don't want extra memory allocations in kernel like fill + conv.

@antiagainst will globalVariable allocate memory or it is just a reference? any suggestion on this?

Reply:
Please do wait for @antiagainst to reply, but if the set and binding numbers overlap, it shouldnt result in a separate allocation I think

Reply:
Global variables with set/binding numbers are just "pointers" to resources allocated by the runtime. So the above semantics description is not exactly accurate. I think I'm missing some contexts here but is this an intermediate state or these two variables pointing to the same resource will persist until the very last? It's okay to have overlapping set/binding numbers in shaders for Vulkan (see the ["Note"](https://renderdoc.org/vkspec_chunked/chap15.html#interfaces-resources-setandbinding) in the spec), but we need to make sure attaching `Aliased` decoration to it if in the very end we actually access the resource with both.

Reply:
Oh, if they are in different functions (which then become different kernel launched) then that's certainly fine.

Reply:
It could happen in the same function too.

Reply:
I think it's still in a same kernel launch, eg

```
module attributes {spv.target_env = #spv.target_env<#spv.vce<v1.3, [Shader], [SPV_KHR_storage_buffer_storage_class]>, {max_compute_workgroup_invocations = 128 : i32, max_compute_workgroup_size = dense<[128, 128, 64]> : vector<3xi32>}>, vkspv.entry_point_schedule = ["ForwardPass_ex_dispatch_8_dispatch_0", "ForwardPass_ex_dispatch_8_dispatch_1"]} {
  func @dispatch_1() attributes {spv.entry_point_abi = {local_size = dense<[32, 1, 1]> : vector<3xi32>}, vkspv.num_workgroups_fn = @ForwardPass_ex_dispatch_8_dispatch_1__num_workgroups__} {
    %0 = iree.placeholder for "interface buffer" {binding = @legacy_io::@arg0, operand_result_index = 0 : i32} : memref<4xf32>
    return
  }
  func @dispatch_0() attributes {spv.entry_point_abi = {local_size = dense<[32, 1, 1]> : vector<3xi32>}, vkspv.num_workgroups_fn = @ForwardPass_ex_dispatch_8_dispatch_0__num_workgroups__} {
    %0 = iree.placeholder for "interface buffer" {binding = @legacy_io::@arg0, operand_result_index = 0 : i32} : memref<1xvector<4xf32>>
    return
  }
  hal.interface @legacy_io attributes {sym_visibility = "private"} {
    hal.interface.binding @arg0, set=0, binding=0, type="StorageBuffer", access="Read"
  }
}
```

How do I add `Aliased` decoration to the spv.globalVariable? I couldn't find a build or method to do it.

Reply:
@hanhanW looks like Alias decorations arent handled correctly in the SPIR-V serialization [here](https://github.com/llvm/llvm-project/blob/7d274aa9bed00cdf1197b2f05140635be90f3362/mlir/lib/Dialect/SPIRV/Serialization/Serializer.cpp#L664). Probably need to add it as a `BoolAttr` with the name being the name of the decoration in camel case, i.e. `aliased` and then teach the `processDecoration` method there to emit the serialization.

Reply:
For your example in the above, it's two entry points, so it's still fine.  In SPIR-V the entry point and all functions/resources statically referenced by it (in a recursive way) should be coherent and pass all validations. "Inconsistencies" across different entry points are okay: a SPIR-V shader module is just a "vehicle" that can host many entry points for different execution models. 

So I think we can move forward and submit this pull request.